src/Workspaces/Core/Portable/CodeStyle/CodeStyleOption.cs

```diff
@@ -3,194 +3,56 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
-using System.Collections.Generic;
 using System.Xml.Linq;
-using Microsoft.CodeAnalysis.Diagnostics;
 
-#if CODE_STYLE
-namespace Microsoft.CodeAnalysis.Internal.Options
-#else
 namespace Microsoft.CodeAnalysis.CodeStyle
-#endif
 {
-    internal interface ICodeStyleOption
-    {
-        XElement ToXElement();
-        object Value { get; }
-        NotificationOption Notification { get; }
-        ICodeStyleOption WithValue(object value);
-        ICodeStyleOption WithNotification(NotificationOption notification);
-    }
-
-    /// <summary>
-    /// Represents a code style option and an associated notification option.  Supports
-    /// being instantiated with T as a <see cref="bool"/> or an <c>enum type</c>.
-    /// 
-    /// CodeStyleOption also has some basic support for migration a <see cref="bool"/> option
-    /// forward to an <c>enum type</c> option.  Specifically, if a previously serialized
-    /// bool-CodeStyleOption is then deserialized into an enum-CodeStyleOption then 'false' 
-    /// values will be migrated to have the 0-value of the enum, and 'true' values will be
-    /// migrated to have the 1-value of the enum.
-    /// 
-    /// Similarly, enum-type code options will serialize out in a way that is compatible with 
-    /// hosts that expect the value to be a boolean.  Specifically, if the enum value is 0 or 1
-    /// then those values will write back as false/true.
-    /// </summary>
+    /// <inheritdoc cref="CodeStyleOption2{T}"/>
     public class CodeStyleOption<T> : ICodeStyleOption, IEquatable<CodeStyleOption<T>>
     {
+        private readonly CodeStyleOption2<T> _codeStyleOptionImpl;
```

Comment:
All the existing code from this type was copied directly into `CodeStyleOption2<T>` being added in shared layer. This public type now just has a handle to the instance of  `CodeStyleOption2<T>` that has all the implementation. Very similar approach has been used for rest of the public types in this PR.